lib/ops/opsservice/operationgroup_test.go

```diff
@@ -288,6 +296,109 @@ func (s *OperationGroupSuite) TestClusterStateModifications(c *check.C) {
 	s.assertServerCount(c, 2)
 }
 
+func (s *OperationGroupSuite) TestCanCreateUpgradeOperation(c *check.C) {
+	// Can't create upgrade operation if cluster isn't active.
+	s.setClusterState(c, ops.SiteStateUpdating)
+	_, err := s.createUpgradeOperation(c, s.clock.Now(), false)
+	c.Assert(err, check.NotNil,
+		check.Commentf("Shouldn't be able to create upgrade operation for non-active cluster"))
+
+	// First upgrade operation should create fine.
+	s.setClusterState(c, ops.SiteStateActive)
+	key, err := s.createUpgradeOperation(c, s.clock.Now().Add(time.Second), false)
+	c.Assert(err, check.IsNil,
+		check.Commentf("Should be able to create first upgrade operation"))
+
+	// Reset the cluster state but the first operation is still in progress.
+	s.setClusterState(c, ops.SiteStateActive)
+	_, err = s.createUpgradeOperation(c, s.clock.Now().Add(2*time.Second), false)
+	c.Assert(err, check.NotNil,
+		check.Commentf("Shouldn't be able to create upgrade operation if another one in progress"))
+
+	// Simulate failed operation and force-reset cluster state.
```

Comment:
nit: Usually it's not recommended to mix multiple tests unless they're similarly structured (eg table-driven tests) - instead, separate each test individually for better overview of what's being tested.

Reply:
This is sort of a single continuous test case where the operation gets created, rolled back, another one created, etc., not separate tests, that's why this is a single test.